imagebuildah/stage_executor.go

```diff
@@ -1521,6 +1590,142 @@ func (s *StageExecutor) tagExistingImage(ctx context.Context, cacheID, output st
+		SignaturePolicyPath: s.executor.signaturePolicyPath,
+		Store:               s.executor.store,
+		SystemContext:       s.executor.systemContext,
+		BlobDirectory:       s.executor.blobDirectory,
+		SignBy:              s.executor.signBy,
+		MaxRetries:          s.executor.maxPullPushRetries,
+		RetryDelay:          s.executor.retryPullPushDelay,
+	}
+	ref, digest, err := buildah.Push(ctx, src, dest, options)
+	if err != nil {
+		return fmt.Errorf("failed pushing cache to %q: %w", dest, err)
+	}
+	logrus.Debugf("successfully pushed cache to dest: %+v with ref:%+v and digest: %v", dest, ref, digest)
+	return nil
+}
+
+// pullCache takes the image source of the cache assuming tag
+// already points to the valid cacheKey and pulls the image to
+// local storage only if it was not already present on local storage
+// or a newer version of cache was found in the upstream repo. If new
+// image was pulled function returns image id otherwise returns empty
+// string "" or error if any error was encontered while pulling the cache.
+func (s *StageExecutor) pullCache(ctx context.Context, src string) (string, error) {
+	logrus.Debugf("trying to pull cache from remote repo: %+v", src)
+	options := buildah.PullOptions{
+		SignaturePolicyPath: s.executor.signaturePolicyPath,
+		Store:               s.executor.store,
+		SystemContext:       s.executor.systemContext,
+		BlobDirectory:       s.executor.blobDirectory,
+		MaxRetries:          s.executor.maxPullPushRetries,
+		RetryDelay:          s.executor.retryPullPushDelay,
+		AllTags:             false,
+		ReportWriter:        nil,
+		PullPolicy:          define.PullIfNewer,
+	}
+	// Get list of images in local storage before we pull image.
+	// We need this to ensure if we actually pulled any image or not
+	// and reason for this is explained below in the same function.
+	//
+	// NOTE, TODO: This logic can be removed once Pull API supports notifying
+	// back if image was actually pulled or not.
+	images, err := s.executor.store.Images()
+	if err != nil {
+		return "", fmt.Errorf("failed while getting images from local storage: %w", err)
+	}
+	id, err := buildah.Pull(ctx, src, options)
+	if err != nil {
+		logrus.Debugf("failed pulling cache from source %s: %v", src, err)
+		return "", fmt.Errorf("failed while pulling cache from %q: %w", src, err)
+	}
+	if id != "" {
+		// Reason for this logic:
+		//
+		// pullCache eventually returns pulled image ID back to the caller
+		// and here buildah will produce a log with image ID notifying to
+		// end-users if anything was actually pulled from remote repo in the
+		// the build output by calling `logCachePulled()`.
+		//
+		// Since following build output is part of stdout lets not return wrong
+		// info and further we are using build output to verify test-cases so
+		// don't return any id, since image was already present on local and buildah
+		// will eventually figure out to use valid cache source from local storage.
```

Comment:
AFAICS we only call this if Buildah _did not_ match any of the local images, so I don’t see any opportunity to “eventually figure out” things.

Reply:
@mtrmac Cache hit attempt happens again here where we still RUN the step but don't commit it https://github.com/containers/buildah/blob/main/imagebuildah/stage_executor.go#L1173 specially happens for cases like `COPY` and `ADD`

Reply:
My mistake, you’re completely right about this path potentially matching. 

But that see the large `addedContentSummary` discussion elsewhere, it would AFAICS never match as is.

I expect we need to have the `addedContentSummary` issue fixed first, and only then we could see the `cachePulled` logic take final shape. (At this point I’d expect for the cache lookup to only happen _once_, in different locations based on the step’s instruction — so the mentioned “eventually future out” would not end up happening.)

Reply:
I think this whole logic can be removed if we optimize cache hits for `COPY` and `ADD` since they always miss the first pass, I created a PR for this: https://github.com/containers/buildah/pull/4128

Reply:
this is already fixed in PR https://github.com/containers/buildah/pull/4128